src/ml/neural_net/style_transfer/mps_style_transfer.m

```diff
@@ -251,6 +271,60 @@ - (instancetype) initWithDev:(id<MTLDevice>) dev
   return self;
 }
 
++ (NSData *) cropImageWidth:(NSData *) data
+                     height:(NSUInteger) height
+                 inputWidth:(NSUInteger) inputWidth
+                outputWidth:(NSUInteger) outputWidth
+                numChannels:(NSUInteger) numChannels {
+  NSUInteger outputLength = height * outputWidth * numChannels;
+  NSMutableData *mutableData = [NSMutableData dataWithLength:outputLength];
+
+  NSUInteger xOffset = inputWidth - outputWidth;
+
+  for (NSUInteger idx = 0; idx < height; idx = idx + 1) {
+    NSUInteger startIndex = idx * inputWidth * numChannels;
+    NSUInteger endIndex = startIndex + (numChannels * (inputWidth - xOffset));
+
+    NSRange range = NSMakeRange(startIndex, endIndex);
+
+    NSUInteger dataOffset = outputWidth * numChannels * idx;
+
+    [data getBytes:mutableData.mutableBytes + dataOffset range:range];
+  }
+
+  return mutableData;
+}
+
++ (NSData *) cropImageHeight:(NSData *) data
+                 inputHeight:(NSUInteger) inputHeight
+                outputHeight:(NSUInteger) outputHeight
+                       width:(NSUInteger) width
+                 numChannels:(NSUInteger) numChannels {
+  NSUInteger outputLength = width * outputHeight * numChannels * sizeof(float);
+  NSMutableData *mutableData = [NSMutableData dataWithLength:outputLength];
+  [data getBytes:mutableData.mutableBytes length:outputLength];
+  return mutableData;
+}
+
++ (NSData *) topLeftCropImage:(NSData *) data
+                  inputHeight:(NSUInteger) inputHeight
+                   inputWidth:(NSUInteger) inputWidth
+                 outputHeight:(NSUInteger) outputHeight
+                  outputWidth:(NSUInteger) outputWidth 
+                  numChannels:(NSUInteger) numChannels {
+  NSData *croppedHeightData = [TCMPSStyleTransfer cropImageHeight:data
+                                                      inputHeight:inputHeight
+                                                     outputHeight:outputHeight
+                                                            width:inputWidth
+                                                      numChannels:numChannels];
+
+  return [TCMPSStyleTransfer cropImageWidth:croppedHeightData
```

Comment:
This incurs two copies of the data. Let’s do it with only one. For each contiguous portion of the output, figure out what to copy from the input. 

For clarity, I recommend specifying the arguments to this function as the desired offset and output height/width

Reply:
I did it with one copy!